desktop/cmp/treemap/TreeMapModel.js

```diff
@@ -0,0 +1,405 @@
+     * @param {string} [c.algorithm] - Layout algorithm to use. Either 'squarified', 'sliceAndDice', 'stripes' or 'strip'.
+     *      Defaults to 'squarified'. @see https://www.highcharts.com/docs/chart-and-series-types/treemap for algorithm examples.
+     * @param {string} [c.colorMode] - Heat color distribution mode. Either 'linear', 'balanced' or 'none'. Defaults to 'linear'.
+     * @param {function} [c.onClick] - Callback to call when a node is clicked. Receives (record, e).
+     *      If not provided, by default will select a record when using a GridModel.
+     * @param {function} [c.onDoubleClick] - Callback to call when a node is double clicked. Receives (record, e).
+     *      If not provided, by default will expand / collapse a record when using a GridModel.
+     * @param {(boolean|TreeMapModel~tooltipFn)} [c.tooltip] - 'true' to use the default tooltip renderer, or a custom
+     *      tooltipFn which returns a string output of the node's value.
+     * @param {(Element|string)} [c.emptyText] - Element/text to render if TreeMap has no records.
+     *      Defaults to null, in which case no empty text will be shown.
+     */
+    constructor({
+        store,
+        gridModel,
+        maxNodes = 1000,
+        highchartsConfig,
+        labelField = 'name',
+        valueField = 'value',
+        heatField = 'value',
+        maxDepth,
+        algorithm = 'squarified',
+        colorMode = 'linear',
+        onClick,
+        onDoubleClick,
+        tooltip = true,
+        emptyText,
+        filter
+    } = {}) {
+        this.gridModel = gridModel;
+        this.store = store ? store : gridModel ? gridModel.store : null;
+        throwIf(!this.store,  'TreeMapModel requires either a Store or a GridModel');
+        this.maxNodes = maxNodes;
+
+        this.highchartsConfig = highchartsConfig;
+        this.labelField = labelField;
+        this.valueField = valueField;
+        this.heatField = heatField;
+        this.maxDepth = maxDepth;
+
+        throwIf(!['sliceAndDice', 'stripes', 'squarified', 'strip'].includes(algorithm), `Algorithm "${algorithm}" not recognised.`);
+        this.algorithm = algorithm;
+
+        throwIf(!['linear', 'balanced', 'none'].includes(colorMode), `Color mode "${colorMode}" not recognised.`);
+        this.colorMode = colorMode;
+
+        this.onClick = withDefault(onClick, this.defaultOnClick);
+        this.onDoubleClick = withDefault(onDoubleClick, this.defaultOnDoubleClick);
+        this.tooltip = tooltip;
+        this.emptyText = emptyText;
+
+        this._filter = filter;
+
+        this.addReaction({
+            track: () => [
+                this.store.rootRecords,
+                this.expandState,
+                this.colorMode,
+                this.labelField,
+                this.valueField,
+                this.heatField,
+                this.maxDepth
+            ],
+            run: ([rawData]) => this.data = this.processData(rawData)
```

Comment:
So interestingly enough, with this setup, we are going to be reprocessing all the data even if we are in error mode?  Its more of a theoretical concern, but does seem wrong that we would still prepare a 5 million node store that we would ultimately not render.  What if our "error" logic was protecting us from other problems that would cause our computation to throw?

A related concern is that we are actually recomputing the color scale whenever we change the expand state.  This seems wrong to me as well....